arrow/src/main/java/org/apache/iceberg/arrow/DictEncodedArrowConverter.java

```diff
@@ -0,0 +1,224 @@
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getLong(idx)));
+    return vector;
+  }
+
+  private static Float4Vector toFloat4Vector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Float4Vector vector =
+        new Float4Vector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getFloat(idx)));
+    return vector;
+  }
+
+  private static Float8Vector toFloat8Vector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Float8Vector vector =
+        new Float8Vector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getDouble(idx)));
+    return vector;
+  }
+
+  private static VarCharVector toVarCharVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    VarCharVector vector =
+        new VarCharVector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(
+        vector,
+        vectorHolder,
+        idx -> vector.set(idx, accessor.getUTF8String(idx).getBytes(StandardCharsets.UTF_8)));
+    return vector;
+  }
+
+  private static VarBinaryVector toVarBinaryVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    VarBinaryVector vector =
+        new VarBinaryVector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getBinary(idx)));
+    return vector;
+  }
+
+  private static TimeMicroVector toTimeMicroVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    TimeMicroVector vector =
+        new TimeMicroVector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getLong(idx)));
+    return vector;
+  }
+
+  private static void initVector(
+      BaseFixedWidthVector vector, VectorHolder vectorHolder, IntConsumer consumer) {
+    vector.allocateNew(vectorHolder.vector().getValueCount());
+    for (int i = 0; i < vectorHolder.vector().getValueCount(); i++) {
+      if (isNullAt(vectorHolder, i)) {
+        vector.setNull(i);
+      } else {
+        consumer.accept(i);
+      }
```

Comment:
Would it make sense to have a common helper here which can be used for both the fixed and variable cases?

```
private static void initVector(FieldVector vec, IntConsumer consumer) {
     if (isNullAt(vectorHolder, i)) {
        vector.setNull(i);
      } else {
        consumer.accept(i);
      }
}
```

Reply:
makes sense, updated